{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b3360614",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "\n",
    "import torch\n",
    "from sklearn import linear_model\n",
    "from sklearn.pipeline import make_pipeline\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "from sklearn.model_selection import StratifiedKFold, cross_validate\n",
    "from sklearn.decomposition import PCA\n",
    "import numpy as np\n",
    "\n",
    "\n",
    "# NOTE:\n",
    "# This notebook assumes execution from the project root.\n",
    "# It explicitly sets cwd to external/genomic-FM so GV-Rep loaders work.\n",
    "\n",
    "os.chdir(\"../external/genomic-FM\")\n",
    "sys.path.append(\".\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e89a347b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dict_keys(['model_name', 'seq_len', 'pooling', 'layer', 'labels', 'ref_embeddings', 'alt_embeddings', 'delta_embeddings', 'ref_sequences', 'alt_sequences'])\n",
      "Embeddings shape: torch.Size([155, 1024])\n",
      "Labels shape: 155\n",
      "Encoded labels: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]\n"
     ]
    }
   ],
   "source": [
    "path = \"root/data/clinvar_embeddings__n155__len1024__layer-1__maskedmean.pt\"\n",
    "payload = torch.load(path, map_location=\"cpu\")\n",
    "\n",
    "print(payload.keys())\n",
    "print(\"Embeddings shape:\", payload[\"delta_embeddings\"].shape)\n",
    "print(\"Labels shape:\", len(payload[\"labels\"]))\n",
    "\n",
    "X = payload[\"delta_embeddings\"]\n",
    "labels = payload[\"labels\"]\n",
    "\n",
    "#Encode labels (Classes 1,2,3 --> 0 and Classes 3,4,5 --> 1)\n",
    "# = [0 if label in [\"Class 1\", \"Class 2\", \"Class 3\"] else 1 for label in labels]\n",
    "\n",
    "mask = np.isin(labels, [\"Class 1\", \"Class 5\"])\n",
    "y = [0 if label == \"Class 1\" else 1 for label in labels if label in [\"Class 1\", \"Class 5\"]]\n",
    "X = X[mask]\n",
    "print(\"Encoded labels:\", y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c4103c7c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ROC-AUC: 0.7253030303030303 +/- 0.044787017665072246\n",
      "Acc:     0.7114624505928855 +/- 0.073479525935951\n",
      "F1:      0.725752508361204 +/- 0.07510328653498402\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/miniconda3/envs/mechanistic_interpretability/lib/python3.11/site-packages/sklearn/linear_model/_logistic.py:1135: FutureWarning: 'penalty' was deprecated in version 1.8 and will be removed in 1.10. To avoid this warning, leave 'penalty' set to its default value and use 'l1_ratio' or 'C' instead. Use l1_ratio=0 instead of penalty='l2', l1_ratio=1 instead of penalty='l1', and C=np.inf instead of penalty=None.\n",
      "  warnings.warn(\n",
      "/opt/miniconda3/envs/mechanistic_interpretability/lib/python3.11/site-packages/sklearn/linear_model/_logistic.py:1135: FutureWarning: 'penalty' was deprecated in version 1.8 and will be removed in 1.10. To avoid this warning, leave 'penalty' set to its default value and use 'l1_ratio' or 'C' instead. Use l1_ratio=0 instead of penalty='l2', l1_ratio=1 instead of penalty='l1', and C=np.inf instead of penalty=None.\n",
      "  warnings.warn(\n",
      "/opt/miniconda3/envs/mechanistic_interpretability/lib/python3.11/site-packages/sklearn/linear_model/_logistic.py:1135: FutureWarning: 'penalty' was deprecated in version 1.8 and will be removed in 1.10. To avoid this warning, leave 'penalty' set to its default value and use 'l1_ratio' or 'C' instead. Use l1_ratio=0 instead of penalty='l2', l1_ratio=1 instead of penalty='l1', and C=np.inf instead of penalty=None.\n",
      "  warnings.warn(\n",
      "/opt/miniconda3/envs/mechanistic_interpretability/lib/python3.11/site-packages/sklearn/linear_model/_logistic.py:1135: FutureWarning: 'penalty' was deprecated in version 1.8 and will be removed in 1.10. To avoid this warning, leave 'penalty' set to its default value and use 'l1_ratio' or 'C' instead. Use l1_ratio=0 instead of penalty='l2', l1_ratio=1 instead of penalty='l1', and C=np.inf instead of penalty=None.\n",
      "  warnings.warn(\n",
      "/opt/miniconda3/envs/mechanistic_interpretability/lib/python3.11/site-packages/sklearn/linear_model/_logistic.py:1135: FutureWarning: 'penalty' was deprecated in version 1.8 and will be removed in 1.10. To avoid this warning, leave 'penalty' set to its default value and use 'l1_ratio' or 'C' instead. Use l1_ratio=0 instead of penalty='l2', l1_ratio=1 instead of penalty='l1', and C=np.inf instead of penalty=None.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "clf = make_pipeline(\n",
    "    StandardScaler(),\n",
    "    PCA(n_components=50),  # optional: reduce dimensionality\n",
    "    LogisticRegression(\n",
    "        penalty=\"l2\",\n",
    "        C=0.1,               # smaller = stronger regularization; start 0.1 or 0.01\n",
    "        solver=\"liblinear\",  # good for small datasets\n",
    "        max_iter=5000\n",
    "    )\n",
    ")\n",
    "\n",
    "cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=0)\n",
    "\n",
    "res = cross_validate(\n",
    "    clf, X, y,\n",
    "    cv=cv,\n",
    "    scoring=[\"roc_auc\", \"accuracy\", \"f1\"],\n",
    "    return_train_score=False\n",
    ")\n",
    "\n",
    "print(\"ROC-AUC:\", res[\"test_roc_auc\"].mean(), \"+/-\", res[\"test_roc_auc\"].std())\n",
    "print(\"Acc:    \", res[\"test_accuracy\"].mean(), \"+/-\", res[\"test_accuracy\"].std())\n",
    "print(\"F1:     \", res[\"test_f1\"].mean(), \"+/-\", res[\"test_f1\"].std())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0860dbd4",
   "metadata": {},
   "source": [
    "This tells us there is enough signal in the embeddings to somewhat discern the two extremes. The nucleotide-transformer last-layer representations encode information that distinguishes benign from pathogenic variants. ClinVar classes dilute separability due to level ambiguity, not model failure.\n",
    "\n",
    "PCA also improves performance a bit, showing how discriminative information lives in a small subspace, and that most dimensions are noise or irrelevant variation. \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5d35747",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mechanistic_interpretability",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
